sw/device/tests/sram_ctrl_execution_test_ret.c

```diff
@@ -0,0 +1,175 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/csr.h"
+#include "sw/device/lib/base/stdasm.h"
+#include "sw/device/lib/dif/dif_sram_ctrl.h"
+#include "sw/device/lib/handler.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/fault_handler_testutils.h"
+#include "sw/device/lib/testing/sram_ctrl_testutils.h"
+#include "sw/device/lib/testing/test_framework/test_main.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+const test_config_t kTestConfig;
+
+static dif_sram_ctrl_t sram_ctrl;
+
+/**
+ * This flag is used to verify that the instruction access fault has occurred,
+ * and has been successfully serviced. It is declared as volatile is referenced
+ * in the fault handler, as well as the main test flow.
+ */
+static volatile bool instruction_access_fault;
+
+static volatile const uint32_t kExecParam1 = 2;
+static volatile const uint32_t kExecParam2 = 3;
+
+static const uint32_t kRamStartAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR;
+static const uint32_t kRamEndAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR +
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_SIZE_BYTES - 1;
+
+/*
+ * Buffer to hold instructions to be executed.
```

Comment:
Why can't this be inline assembly inside of a naked function?
```c
OT_ATTR_NAKED
OT_ATTR_SECTION(".data")
void do_test_things(void) {
  asm(...);
}
```
The `.data` bit will force the function to be placed into ram for you by CRT initialization.
(You may need to go define `OT_ATTR_NAKED` in macros.h)

Reply:
Can you use inline assembly to declare the instructions anyways? I am not comfortable approving  a change with hand-encoded instructions.

Reply:
Not opposed to it, but not sure how I force inline assembly into a buffer. If it was `.data` segment, your example above looks good.

I suppose as the function is naked, and we know that there are only two instructions - we could then just copy the code into Retention SRAM, assuming 64bytes. But is it really that much better than two handcrafted instructions that are will documented?

Reply:
I think that hardcoding the instructions is reasonable here. The comments do a good job of explaining what is going on. As @silvestrst says copying the function is not foolproof either.